server/src/main/java/org/elasticsearch/env/NodeMetadata.java

```diff
@@ -145,8 +167,13 @@ public NodeMetadata build() {
             if (this.previousNodeVersion == null) {
                 previousNodeVersion = nodeVersion;
             }
+            if (this.oldestIndexVersion == null) {
+                oldestIndexVersion = Version.V_EMPTY;
```

Comment:
This is only used in the legacy format which we shouldn't really be reading here, but I think this means that the index compatibility check will fail before the node version check happens, so the error will say that there are some ancient indices rather than that the previous node version is too old. I wonder if it would be better to check the index version in `NodeMetadata#upgradeToCurrentVersion` and just call that in both places so we have consistent checks.

Reply:
Yes exactly. I was testing direct to 8 upgrade from 6.8 and this showed up. Right now it says you have indices from version 0.0.0, which does the job but isn't the best. Good suggestion, I'll extract the upgrade/downgrade check on the metadata in a separate method and call that ahead.